### PR TITLE
e2e topology manager: single-numa-node multi container tests

### PR DESCRIPTION
**Pull request by ffromani**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This is a followup of https://github.com/kubernetes/kubernetes/pull/87645 . I'm working to add more e2e tests for the topology manager
to fix the issue https://github.com/kubernetes/kubernetes/issues/83481
This PR adds minor fixes and improvements that didn't make it into https://github.com/kubernetes/kubernetes/pull/87645 and adds e2e node tests for multi-container pods.

**Which issue(s) this PR fixes**:
Related to https://github.com/kubernetes/kubernetes/issues/83481

**Special notes for your reviewer**:
N/A

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by fromanirh** [comment]
/hold
The PR is reviewable from my perspective but the final round of testing is still in progress

**Comment by fromanirh** [comment]
bad rebase, missing one patch which causes a false negative. Fixing ASAP.

**Comment by fromanirh** [comment]
/hold cancel
all tests succesful on my side

**Comment by fromanirh** [comment]
i signed it

**Comment by fromanirh** [comment]
force-pushed the series because the -verify job highlighted some issues with patches early in the series.

File: test/e2e_node/topology_manager_test.go [hunk 1]
```diff
@@ -598,7 +627,16 @@ func getSRIOVDevicePluginConfigMap(cmFile string) *v1.ConfigMap {
...
+type sriovData struct {
+	configMap      *v1.ConfigMap
+	serviceAccount *v1.ServiceAccount
+	pod            *v1.Pod
+
+	resourceName   string
+	resourceAmount int64
+}
+
+func setupSRIOVConfigOrFail(f *framework.Framework, configMap *v1.ConfigMap) sriovData {
```
Comment (cynepco3hahue) [inline test/e2e_node/topology_manager_test.go]: let's avoid the copy of the struct so return *sriovData

File: test/e2e_node/numa_alignment.go [hunk 2]
```diff
@@ -184,22 +177,35 @@ func makeEnvMap(logs string) (map[string]string, error) {
 	return envMap, nil
 }
 
-func checkNUMAAlignment(f *framework.Framework, pod *v1.Pod, logs string, numaNodes int) (numaPodResources, error) {
+type testEnvInfo struct {
+	numaNodes         int
+	sriovResourceName string
+}
+
+func containerWantsDevices(cnt *v1.Container, envInfo testEnvInfo) bool {
```
Comment (cynepco3hahue) [inline test/e2e_node/numa_alignment.go]: pass the pointer on `testEnvInfo`, the same true for all appearances of `testEnvInfo`

File: test/e2e_node/numa_alignment.go [hunk 3]
```diff
@@ -184,22 +177,35 @@ func makeEnvMap(logs string) (map[string]string, error) {
...
+	_, found := cnt.Resources.Requests[v1.ResourceName(envInfo.sriovResourceName)]
+	return found
+}
+
+func checkNUMAAlignment(f *framework.Framework, pod *v1.Pod, cnt *v1.Container, logs string, envInfo testEnvInfo) (numaPodResources, error) {
```
Comment (cynepco3hahue) [inline test/e2e_node/numa_alignment.go]: return pointer on `numaPodResources`, the same true for all appearances of `numaPodResources`
and you can return `return nil, err`

File: test/e2e_node/topology_manager_test.go [hunk 4]
```diff
@@ -255,25 +297,17 @@ func findSRIOVResource(node *v1.Node) (string, int64) {
...
 
-func deletePodInNamespace(f *framework.Framework, namespace, name string) {
-	gp := int64(0)
-	deleteOptions := metav1.DeleteOptions{
-		GracePeriodSeconds: &gp,
-	}
-	err := f.ClientSet.CoreV1().Pods(namespace).Delete(context.TODO(), name, &deleteOptions)
-	framework.ExpectNoError(err)
-}
+func validatePodAlignment(f *framework.Framework, pod *v1.Pod, envInfo testEnvInfo) {
```
Comment (cynepco3hahue) [inline test/e2e_node/topology_manager_test.go]: pass pointer on `testEnvInfo`

File: test/e2e_node/topology_manager_test.go [hunk 5]
```diff
@@ -502,21 +536,27 @@ func runTopologyManagerPolicySuiteTests(f *framework.Framework) {
...
 			},
+		})
+		if err != nil {
+			return false
 		}
+		return len(containers) == 0
+	}, 2*time.Minute, 1*time.Second).Should(gomega.BeTrue())
+}
+
+func runTopologyManagerPositiveTest(f *framework.Framework, numPods int, ctnAttrs []tmCtnAttribute, envInfo testEnvInfo) {
```
Comment (cynepco3hahue) [inline test/e2e_node/topology_manager_test.go]: pass pointer on `testEnvInfo`

File: test/e2e_node/topology_manager_test.go [hunk 6]
```diff
@@ -632,14 +670,34 @@ func setupSRIOVConfigOrFail(f *framework.Framework, configMap *v1.ConfigMap) (*v
...
+		resourceAmount: sriovResourceAmount,
+	}
 }
 
-func teardownSRIOVConfigOrFail(f *framework.Framework, dpPod *v1.Pod) {
-	framework.Logf("deleting the SRIOV device plugin pod %s/%s and waiting for container %s removal",
-		dpPod.Namespace, dpPod.Name, dpPod.Spec.Containers[0].Name)
-	deletePodInNamespace(f, dpPod.Namespace, dpPod.Name)
-	waitForContainerRemoval(dpPod.Spec.Containers[0].Name, dpPod.Name, dpPod.Namespace)
+func teardownSRIOVConfigOrFail(f *framework.Framework, sd sriovData) {
```
Comment (cynepco3hahue) [inline test/e2e_node/topology_manager_test.go]: pass pointer on `sriovData`

File: test/e2e_node/topology_manager_test.go [hunk 7]
```diff
@@ -648,43 +706,173 @@ func runTopologyManagerNodeAlignmentSuiteTests(f *framework.Framework, configMap
...
+	if sd.resourceAmount > 1 {
 		// no matter how busses are connected to NUMA nodes and SRIOV devices are installed, this function
 		// preconditions must ensure the following can be fulfilled
-		ginkgo.By(fmt.Sprintf("Successfully admit two guaranteed pods, each with 1 core, 1 %s device", sriovResourceName))
-		runTopologyManagerPositiveTest(f, numaNodes, 2, "1000m", sriovResourceName, "1")
+		ginkgo.By(fmt.Sprintf("Successfully admit two guaranteed pods, each with 1 core, 1 %s device", sd.resourceName))
+		ctnAttrs = []tmCtnAttribute{
+			{
+				ctnName:       "gu-container",
+				cpuRequest:    "1000m",
```
Comment (cynepco3hahue) [inline test/e2e_node/topology_manager_test.go]: Just curious, does CPU manager returns correct topology hints when you do not specify integer CPU's, I was sure it works only for static policy.

Comment (fromanirh) [reply test/e2e_node/topology_manager_test.go]: In the `configureTopologyManagerInKubelet` helper we always set the CPUManager policy to `static` indeed.

Comment (cynepco3hahue) [reply test/e2e_node/topology_manager_test.go]: Thanks for the clarification.

**Comment by cynepco3hahue** [comment]
lgtm

**Comment by derekwaynecarr** [comment]
the tests look good, appreciate the coverage with both positive and negative scenarios.

/lgtm
/approve

**Pull request merged**